From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bacteriawa <A3167717663@hotmail.com>
Date: Sun, 6 Jul 2025 03:03:12 +0800
Subject: [PATCH] Rebrand to XenonFolia


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index db5e2f10278c20286779362cfdfadb2f42c724e0..b97000999753355f6a5b59f76521cccf06c71b1e 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -28,6 +28,10 @@ public interface ServerBuildInfo {
      * The brand id for Luminol.
      */
     Key BRAND_LUMINOL_ID = Key.key("luminolmc", "luminol");
+    /**
+     * The brand id for XenonFolia.
+     */
+    Key BRAND_XENONFOLIA_ID = Key.key("xenonmc", "xenonfolia");
     // Luminol end
 
     /**
